### ISSUE TITLE
Encoder JSON from protobuf generate Timestamp with negative hours

### ISSUE THREAD

**Issue by bernardo-bruning** [protocolbuffers/protobuf#9424]
Version: v3.19..3
Language: Ruby
Operation system: Linux
ruby 2.7.2

Currently, when I try encoding to JSON  a Timestamp less than 1970 the protobuf produces timestamp with negative hours.

See steps:

```irb
irb(main):014:0> require 'google/protobuf'
=> true
irb(main):015:0> require 'google/protobuf/timestamp_pb'
=> true
irb(main):016:0> 
irb(main):017:0> time = Time.new(1900, 01, 01, 03, 00, 00)
=> 1900-01-01 03:00:00 -0313
irb(main):018:0> time_protobuf = Google::Protobuf::Timestamp.new(seconds: time.to_i, nanos: time.nsec)
=> <Google::Protobuf::Timestamp: seconds: -2208966364, nanos: 0>
irb(main):019:0> time_protobuf.to_json
=> "\"1900-01-02T-17:-46:-4Z\""
irb(main):020:0> 
irb(main):021:0> # If tries decode the JSON raises an exception
=> nil
irb(main):022:0> Google::Protobuf::Timestamp.decode_json(time_protobuf.to_json)
Traceback (most recent call last):
        5: from /home/bernardobruning/.rbenv/versions/2.7.2/bin/irb:23:in `<main>'
        4: from /home/bernardobruning/.rbenv/versions/2.7.2/bin/irb:23:in `load'
        3: from /home/bernardobruning/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/irb-1.2.6/exe/irb:11:in `<top (required)>'
        2: from (irb):22
        1: from (irb):22:in `decode_json'
Google::Protobuf::ParseError (Error occurred during parsing: Error parsing JSON @1:24: Malformed timestamp)
```

The result expected is not negative hours, but is encoded to  `1900-01-02T-17:-46:-4Z`. Is expected to decode the JSON that protobuf encoded.

**Comment by bernardo-bruning** [comment]
Hi, I have been working on a fix to this issue. I resolved negative hours but kept working to fix the day calculation. I believe that I can assign myself to this issue and believe that problems also occur in PHP because the code is similar.

Ruby [code](https://github.com/protocolbuffers/protobuf/blob/66d749188ff2a2e30e932110222d58da7c6a8d49/php/ext/google/protobuf/php-upb.c#L8907).
PHP [code](https://github.com/protocolbuffers/protobuf/blob/66d749188ff2a2e30e932110222d58da7c6a8d49/ruby/ext/google/protobuf_c/ruby-upb.c).

**Comment by haberman** [comment]
Good catch. The place to make the fix is in the upb repo where the code actually lives:

https://github.com/protocolbuffers/upb/blob/85e5c76cb7098d3ae35e575d652a40568013a149/upb/json_encode.c#L142-L183

**Comment by bernardo-bruning** [comment]
Thanks! It's automatically replicated to here?

**Comment by haberman** [comment]
The change will propagate here with https://github.com/protocolbuffers/protobuf/pull/9589, which should land before the 3.20 release.

**Comment by haberman** [comment]
This should have landed in 3.20.0. Please re-open if you still experience this error.
